pkg/controller/rollout/rollout_canary.go

```diff
@@ -0,0 +1,504 @@
+}
+
+func (m *canaryReleaseManager) fetchBatchRelease(ns, name string) (*v1alpha1.BatchRelease, error) {
+	br := &v1alpha1.BatchRelease{}
+	// batchRelease.name is equal related rollout.name
+	err := m.Get(context.TODO(), client.ObjectKey{Namespace: ns, Name: name}, br)
+	return br, err
+}
+
+func createBatchRelease(rollout *v1alpha1.Rollout, rolloutID string, batch int32, isRollback bool) *v1alpha1.BatchRelease {
+	var batches []v1alpha1.ReleaseBatch
+	for _, step := range rollout.Spec.Strategy.Canary.Steps {
+		if step.Replicas == nil {
+			batches = append(batches, v1alpha1.ReleaseBatch{CanaryReplicas: intstr.FromString(strconv.Itoa(int(*step.Weight)) + "%")})
+		} else {
+			batches = append(batches, v1alpha1.ReleaseBatch{CanaryReplicas: *step.Replicas})
+		}
+	}
+	br := &v1alpha1.BatchRelease{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:       rollout.Namespace,
+			Name:            rollout.Name,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(rollout, rolloutControllerKind)},
+		},
+		Spec: v1alpha1.BatchReleaseSpec{
+			TargetRef: v1alpha1.ObjectRef{
+				WorkloadRef: &v1alpha1.WorkloadRef{
+					APIVersion: rollout.Spec.ObjectRef.WorkloadRef.APIVersion,
+					Kind:       rollout.Spec.ObjectRef.WorkloadRef.Kind,
+					Name:       rollout.Spec.ObjectRef.WorkloadRef.Name,
+				},
+			},
+			ReleasePlan: v1alpha1.ReleasePlan{
+				Batches:          batches,
+				RolloutID:        rolloutID,
+				BatchPartition:   utilpointer.Int32Ptr(batch),
+				FailureThreshold: rollout.Spec.Strategy.Canary.FailureThreshold,
+			},
+		},
+	}
+	if isRollback {
+		if br.Annotations == nil {
+			br.Annotations = map[string]string{}
+		}
+		br.Annotations[v1alpha1.RollbackInBatchAnnotation] = "true"
+	}
+	return br
+}
+
+func (m *canaryReleaseManager) removeBatchRelease(c *util.RolloutContext) (bool, error) {
+	batch := &v1alpha1.BatchRelease{}
+	err := m.Get(context.TODO(), client.ObjectKey{Namespace: c.Rollout.Namespace, Name: c.Rollout.Name}, batch)
+	if err != nil && errors.IsNotFound(err) {
+		return true, nil
+	} else if err != nil {
+		klog.Errorf("rollout(%s/%s) fetch BatchRelease failed: %s", c.Rollout.Namespace, c.Rollout.Name)
+		return false, err
+	}
+	if !batch.DeletionTimestamp.IsZero() {
+		klog.Infof("rollout(%s/%s) BatchRelease is terminating, and wait a moment", c.Rollout.Namespace, c.Rollout.Name)
+		return false, nil
+	}
+
+	//delete batchRelease
+	err = m.Delete(context.TODO(), batch)
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) delete BatchRelease failed: %s", c.Rollout.Namespace, c.Rollout.Name, err.Error())
+		return false, err
+	}
+	klog.Infof("rollout(%s/%s) deleting BatchRelease, and wait a moment", c.Rollout.Namespace, c.Rollout.Name)
+	return false, nil
+}
+
+// todo, remove the section, and batchRelease will implement the process
+// 1. set workload.pause = false
+// 2. set workload.partition = 0
+func (m *canaryReleaseManager) restoreWorkload(c *util.RolloutContext) (bool, error) {
```

Comment:
you cloud replace the func with  [this](https://github.com/openkruise/rollouts/blob/c0b1fea7f8e39e89d4fa544ed2e43ae6e291e50e/pkg/controller/rollout/batchrelease/inner_batchrelease.go#L187) .